README.md

```diff
@@ -66,55 +66,65 @@ flowchart LR
 
     Library -->|Call C code| libddwaf
 
     classDef hidden display: none;
 ```
 
-### Allocator
+### CGO Reference Pool
 
-The cgoRefPool is a pure Go cgoRefPool of `ddwaf_object` C values on the Go memory heap.
-the `cgoRefPool` go type is a way to make sure we can safely send go allocated data to the C side of the WAF
+The cgoRefPool type is a pure Go pointer pool of `ddwaf_object` C values on the Go memory heap.
+the `cgoRefPool` go type is a way to make sure we can safely send Go allocated data to the C side of the WAF
 The main issue is the following: the `wafObject` uses a C union to store the tree structure of the full object,
 union equivalent in go are interfaces and they are not compatible with C unions. The only way to be 100% sure
 that the Go `wafObject` struct has the same layout as the C one is to only use primitive types. So the only way to
 store a raw pointer is to use the `uintptr` type. But since `uintptr` do not have pointer semantics (and are just
-basically integers), we need another structure to store the value as Go pointer because the GC is lurking. That's
-where the `cgoRefPool` object comes into play: all new `wafObject` elements are created via this API whose especially
+basically integers), we need another method to store the value as Go pointer because the GC will delete our data if it
+not reference bu Go pointers.
+
+That's where the `cgoRefPool` object comes into play: all new `wafObject` elements are created via this API whose especially
 built to make sure there is no gap for the Garbage Collector to exploit. From there, since underlying values of the
-`wafObject` are either arrays (for maps, structs and arrays) or string (for all ints, booleans and strings),
+`wafObject` are either arrays of wafObjects (for maps, structs and arrays) or string (for all ints, booleans and strings),
 we can store 2 slices of arrays and use `runtime.KeepAlive` in each code path to protect them from the GC.
 
+All these objects stored in the reference pool need to live throughout the use of the Waf Context associated.
+
 ### Typical call to Run()
 
 Here is an example of the flow of operations on a simple call to Run():
 
-- Encode input data into Waf Objects
+- Encode input data into Waf Objects and store references in the temporary pool
 - Lock the context mutex until the end of the call
+- store references from the temporary pool into the context level pool
 - Call `ddwaf_run`
 - Decode the matches and actions
 
 ### CGO-less C Bindings
 
-The main component used to build C bindings without using CGO is called [purego](https://github.com/ebitengine/purego). The flow of execution on our side is to embed the C shared library using `go:embed`. Then to dump it into a file, load it using `dlopen` and to load the symbols using `dlsym`. And finally to call them.
+The main component used to build C bindings without using CGO is called [purego](https://github.com/ebitengine/purego). The flow of execution on our side
```

Comment:
I find `build` a bit overloaded here especially as there are mentions of `cgo` here... I also find a more clinical phrasing is clearer (feel free to disagree):

```suggestion
This library uses [purego](https://github.com/ebitengine/purego) to implement C bindings without requiring use of CGO at compilation time. The high-level workflow
```